OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -434,15 +437,28 @@ public virtual int GetBuildTime(ActorInfo unit, BuildableInfo bi)
 
 			var time = bi.BuildDuration;
 			if (time == -1)
-			{
-				var valued = unit.TraitInfoOrDefault<ValuedInfo>();
-				time = valued != null ? valued.Cost : 0;
-			}
+				time = GetProductionCost(unit);
+
+			var modifiers = unit.TraitInfos<IProductionTimeModifierInfo>().Select(t => t.GetProductionTimeModifier(techTree, Info.Type));
+			time = Util.ApplyPercentageModifiers(time, modifiers);
```

Comment:
This can simplify to
```csharp
var modifiers = unit.TraitInfos<IProductionTimeModifierInfo>()
	.Select(t => t.GetProductionTimeModifier(techTree, Info.Type))
	.Append(bi.BuildDurationModifier)
	.Append(Info.BuildDurationModifier);

return Util.ApplyPercentageModifiers(time, modifiers);
```